### PR TITLE
fix-NXP-14830-detect-group-changes

### PR DESCRIPTION
**Pull request by ataillefer**

### CODE REVIEW

File: nuxeo-services/nuxeo-platform-usermanager-core/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java [hunk 1]
```diff
@@ -114,7 +114,20 @@
...
+
+    /**
+     * @since 9.2
+     */
+    public static final String ID_PROPERTY_KEY = "id";
+
+    /**
+     * @since 9.2
+     */
+    public static final String ANCESTOR_GROUPS_PROPERTY_KEY = "ancestorGroups";
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-usermanager-core/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java]: Please explain in the Javadoc that these constants are keys in the event context.

File: nuxeo-services/nuxeo-platform-usermanager-core/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java [hunk 2]
```diff
@@ -616,11 +616,15 @@ public void testGetUsersInGroupAndSubgroupsWithoutInfiniteLoop() throws Exceptio
...
-        // infinite loop can occure here:
+        // infinite loop can occur here:
         List<String> usersInGroupAndSubGroups2 = userManager.getUsersInGroupAndSubGroups("test_g2");
         Collections.sort(expectedUsersInGroup2);
         Collections.sort(usersInGroupAndSubGroups2);
         assertEquals(expectedUsersInGroup2, usersInGroupAndSubGroups2);
+
+        // and here
+        List<String> g1AncestorGroups = userManager.getAncestorGroups("test_g1");
+        assertEquals(Arrays.asList("test_g2", "test_g1"), g1AncestorGroups);
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-usermanager-core/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java]: There's no order guarantee. Use a Set for comparison, or sort the list.

File: nuxeo-services/nuxeo-platform-usermanager-core/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java [hunk 3]
```diff
@@ -1156,4 +1160,48 @@ protected Object resolveObject(Object obj) throws IOException {
...
+
+        DocumentModel groupC = getGroup("C");
+        groupC.setPropertyValue("group:subGroups", (Serializable) Collections.singletonList("CD"));
+        userManager.createGroup(groupC);
+
+        DocumentModel groupD = getGroup("D");
+        groupD.setPropertyValue("group:subGroups", (Serializable) Collections.singletonList("CD"));
+        userManager.createGroup(groupD);
+
+        assertEquals(Arrays.asList("AB", "A", "B", "CD", "C", "D"), userManager.getAncestorGroups("ABCD"));
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-usermanager-core/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java]: There's no order guarantee. Use a Set for comparison, or sort the lists.

File: nuxeo-services/nuxeo-platform-usermanager-core/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java [hunk 4]
```diff
@@ -748,13 +761,23 @@ public String getUserSortField() {
      * @since 8.2
      */
     private void notifyCore(String userOrGroupId, String eventId) {
+        notifyCore(userOrGroupId, eventId, null);
+    }
+
+    /**
+     * @since 9.2
+     */
+    private void notifyCore(String userOrGroupId, String eventId, List<String> ancestorGroupIds) {
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-usermanager-core/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java]: `private` -> `protected`

File: nuxeo-services/nuxeo-platform-usermanager-core/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java (hunk 2 above)
Comment (ataillefer) [inline nuxeo-services/nuxeo-platform-usermanager-core/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java]: In fact there is due to the implementation but yes I will.

**Comment by efge** [comment]
👍

**Comment by kevinleturc** [comment]
👍

**Pull request merged**